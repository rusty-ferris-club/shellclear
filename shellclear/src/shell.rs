use std::{fmt, path::Path};

use serde_derive::{Deserialize, Serialize};
use strum::{EnumIter, IntoEnumIterator};

/// List of all supported shells
#[derive(Debug, EnumIter, Clone, Eq, Hash, PartialEq)]
pub enum Shell {
    Bash,
    Zshrc,
    Fish,
    PowerShell,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct FishHistory {
    pub cmd: String,
    pub when: String,
}

impl fmt::Display for Shell {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Self::Bash => write!(f, "bash"),
            Self::Zshrc => write!(f, "zshrc"),
            Self::Fish => write!(f, "fish"),
            Self::PowerShell => write!(f, "powershell"),
        }
    }
}

/// Zsh history file name
const ZSH_HISTORY_FILE_PATH: &str = ".zsh_history";
/// Bash history file name
const BASH_HISTORY_FILE_PATH: &str = ".bash_history";
const FISH_HISTORY_FILE_PATH: &str = ".local/share/fish/fish_history";
const POWERSHELL_HISTORY_FILE_PATH: &str =
    "AppData/Roaming/Microsoft/Windows/PowerShell/PSReadline/ConsoleHost_history.txt";

/// History shell details
#[derive(Clone, Debug)]
pub struct History {
    /// shell type
    pub shell: Shell,
    /// history file path
    pub path: String,
    /// history file name
    pub file_name: String,
}

/// return list of all existing history files
#[must_use]
pub fn get_all_history_files(homedir: &str) -> Vec<History> {
    // return list of existing shells
    Shell::iter()
        .filter_map(|shell| {
            let shell_history_path = get_shell_history_path(&shell, homedir);
            if !Path::new(&shell_history_path).exists() {
                log::debug!("shell {:?} not found", shell);
                return None;
            }
            let file_name = Path::new(&shell_history_path)
                .file_name()?
                .to_str()?
                .to_string();
            Some(History {
                shell,
                path: shell_history_path,
                file_name,
            })
        })
        .collect::<Vec<_>>()
}

// returns all supported shells types
fn get_shell_history_path(shell_type: &Shell, homedir: &str) -> String {
    match shell_type {
        Shell::Bash => Path::new(homedir)
            .join(BASH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::Zshrc => Path::new(homedir)
            .join(ZSH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::Fish => Path::new(homedir)
            .join(FISH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::PowerShell => Path::new(homedir)
            .join(POWERSHELL_HISTORY_FILE_PATH)
            .display()
            .to_string(),
    }
}

#[cfg(test)]
mod state_shell {
    use std::{fs, fs::File};

    use insta::{assert_debug_snapshot, with_settings};
    use tempdir::TempDir;

    use super::{
        get_all_history_files, BASH_HISTORY_FILE_PATH, FISH_HISTORY_FILE_PATH,
        ZSH_HISTORY_FILE_PATH,
    };

    #[test]
    fn can_get_all_history_files() {
        let temp_dir = TempDir::new("terminal").unwrap().path().join("app");

        fs::create_dir_all(&temp_dir).unwrap();
        fs::create_dir_all(temp_dir.join(FISH_HISTORY_FILE_PATH).parent().unwrap()).unwrap();

        File::create(temp_dir.join(ZSH_HISTORY_FILE_PATH)).expect("create failed");
        File::create(temp_dir.join(BASH_HISTORY_FILE_PATH)).expect("create failed");
        File::create(temp_dir.join(FISH_HISTORY_FILE_PATH)).expect("create failed");

        with_settings!({filters => vec![
            (r"//*.+/(app)", "PATH"),
            (r"([C]?:\\.+app\\\\)", "PATH/")// for windows
        ]}, {
            assert_debug_snapshot!(get_all_history_files(&temp_dir.display().to_string()));
        });
    }
}
