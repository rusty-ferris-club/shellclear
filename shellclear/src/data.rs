use serde_derive::Deserialize;

use crate::shell::Shell;

pub const ROOT_APP_FOLDER: &str = env!("CARGO_PKG_NAME");

#[derive(Debug)]
pub struct CmdExit {
    pub code: exitcode::ExitCode,
    pub message: Option<String>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct SensitiveCommands {
    #[serde(with = "serde_regex")]
    pub test: regex::Regex,
    pub name: String,
    pub secret_group: u8,
    #[serde(default)]
    pub id: String,
}

#[derive(Debug, Clone)]
pub struct FindingSensitiveCommands {
    pub shell_type: Shell,
    pub sensitive_findings: Vec<SensitiveCommands>,
    pub command: String,
    pub data: String,
    pub secrets: Vec<String>,
}
