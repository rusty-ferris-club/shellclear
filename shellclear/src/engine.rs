use std::{
    fmt::Write,
    fs::{write, File},
    io::{prelude::*, BufReader},
    time::Instant,
};
use std::collections::HashMap;

use anyhow::Result;
use log::debug;
use rayon::prelude::*;
use regex::Match;


// TODO: Remove clear param
// TODO: Tests
// TODO: comments

use crate::{
    config::Config,
    data::{FindingSensitiveCommands, SensitiveCommands},
    shell,
    state::ShellContext,
    masker::Masker,
};
use crate::shell::Shell;

pub const SENSITIVE_COMMANDS: &str = include_str!("sensitive-patterns.yaml");

pub struct PatternsEngine {
    commands: Vec<SensitiveCommands>,
    masker: Masker,
}

#[derive(Default, Debug)]
pub struct Findings {
    pub patterns: HashMap<Shell, Vec<FindingSensitiveCommands>>,
}

impl Findings {
    #[must_use]
    // add list of finding
    pub fn add_findings(mut self, shell_type: &Shell, finding: Vec<FindingSensitiveCommands>) -> Self {
        if let Some(vec) = self.patterns.get_mut(&shell_type) {
            vec.extend(finding);
        } else {
            self.patterns.insert(shell_type.clone(), finding);
        }

        self
    }

    #[must_use]
    // return list of sensitive findings commands by shell
    pub fn get_sensitive_commands(&self, shell_type: &Shell) -> Vec<&FindingSensitiveCommands> {
        if let Some(vec) = self.patterns.get(shell_type) {
            return vec.iter()
                .filter(|&f| !f.sensitive_findings.is_empty())
                .collect::<Vec<_>>();
        }

        vec![]
    }

    // return list of all sensitive findings commands
    pub fn get_all_sensitive_commands(&self) -> Vec<&FindingSensitiveCommands> {
        self.patterns
            .values()
            .flatten()
            .filter(|&f| !f.sensitive_findings.is_empty())
            .collect::<Vec<_>>()
    }
}

impl PatternsEngine {
    /// Load the engine with config
    ///
    /// # Errors
    ///
    /// Will return `Err` when could not load default sensitive commands
    pub fn with_config(config: &Config) -> Result<Self> {
        let sensitive_patterns = {
            let mut patterns: Vec<SensitiveCommands> = serde_yaml::from_str(SENSITIVE_COMMANDS)?;

            patterns = if config.is_app_path_exists() {
                // load external patterns
                match config.load_patterns_from_default_path() {
                    Ok(p) => patterns.extend(p),
                    Err(e) => log::debug!("could not load external pattern. {:?}", e),
                };

                // ignore patterns
                match config.get_ignore_patterns() {
                    Ok(ignores) => patterns
                        .iter()
                        .filter(|p| !ignores.contains(&p.id))
                        .cloned()
                        .collect::<Vec<_>>(),
                    Err(e) => {
                        log::debug!("could not load ignore pattern. {:?}", e);
                        patterns
                    }
                }
            } else {
                log::debug!(
                    "app config folder not found in path: {}",
                    &config.app_path.display()
                );
                patterns
            };

            patterns
        };
        Ok(Self {
            commands: sensitive_patterns,
            masker: Masker::new(),
        })
    }
    /// Search sensitive command patterns from the given shell list
    ///
    /// # Errors
    ///
    /// Will return `Err` when has an error when find sensitive patterns in a
    /// specific shell
    pub fn find_history_commands_from_shell_list(
        &self,
        shells_context: &Vec<ShellContext>,
    ) -> Result<Findings> {
        let mut findings = Findings::default();

        for shell_context in shells_context {
            findings = findings.add_findings(&shell_context.history.shell, self.find_history_commands(shell_context)?);
        }
        Ok(findings)
    }

    /// Search sensitive command patterns
    ///
    /// # Errors
    ///
    /// Will return `Err` if history file not exists/ couldn't open
    pub fn find_history_commands(
        &self,
        state_context: &ShellContext,
    ) -> Result<Vec<FindingSensitiveCommands>> {
        debug!(
            "clear history commands from path: {}",
            state_context.history.path
        );

        match state_context.history.shell {
            shell::Shell::Fish => self.find_fish(state_context, &self.commands),
            _ => self.find_by_lines(state_context, &self.commands),
        }
    }

    fn find_by_lines(
        &self,
        state_context: &ShellContext,
        sensitive_commands: &[SensitiveCommands],
    ) -> Result<Vec<FindingSensitiveCommands>> {
        let file = File::open(&state_context.history.path)?;
        let reader = BufReader::new(file);

        let start = Instant::now();

        let lines = reader
            .lines()
            .filter(std::result::Result::is_ok)
            .map(std::result::Result::unwrap)
            .collect::<Vec<_>>();

        debug!(
            "time elapsed to read history file: {:?}. found {} commands",
            start.elapsed(),
            lines.len()
        );

        let start = Instant::now();
        let mut results = lines
            .par_iter()
            .map(|command| {
                let (secrets, sensitive_findings) = self.find_secrets(command, sensitive_commands);

                let only_command = match command.split_once(';') {
                    Some((_x, y)) => y.to_string(),
                    _ => command.clone(),
                };

                FindingSensitiveCommands {
                    shell_type: state_context.history.shell.clone(),
                    sensitive_findings,
                    command: only_command,
                    data: command.clone(),
                    secrets,
                }
            })
            .collect::<Vec<_>>();

        self.masker.mask_sensitive_findings(results.as_mut());

        debug!(
            "time elapsed for detect sensitive commands: {:?}",
            start.elapsed()
        );

        Ok(results)
    }

    fn find_fish(
        &self,
        state_context: &ShellContext,
        sensitive_commands: &[SensitiveCommands],
    ) -> Result<Vec<FindingSensitiveCommands>> {
        let start = Instant::now();
        let history: Vec<shell::FishHistory> =
            serde_yaml::from_reader(File::open(&state_context.history.path)?)?;

        let mut results = history
            .par_iter()
            .map(|h| {
                let (secrets, sensitive_findings) = self.find_secrets(&h.cmd, sensitive_commands);

                FindingSensitiveCommands {
                    shell_type: state_context.history.shell.clone(),
                    sensitive_findings,
                    command: h.cmd.clone(),
                    data: serde_yaml::to_string(&h).unwrap(),
                    secrets,
                }
            })
            .collect::<Vec<_>>();

        self.masker.mask_sensitive_findings(results.as_mut());

        debug!(
            "time elapsed to read history file: {:?}. found {} commands",
            start.elapsed(),
            history.len()
        );

        Ok(results)
    }

    fn find_secrets(&self, command: &str, sensitive_commands: &[SensitiveCommands]) -> (Vec<String>, Vec<SensitiveCommands>) {
        let (secrets, sensitive_findings): (Vec<Option<Match>>, Vec<SensitiveCommands>) = sensitive_commands
            .par_iter()
            .filter_map(|v| {
                let capture = v.test.captures(command)?;

                Some((capture.get(v.secret_group as usize), v.clone()))
            })
            .unzip();

        let secrets = secrets.iter()
            .flatten()
            .map(|m| m.as_str().to_string())
            .collect();

        (secrets, sensitive_findings)
    }
}

#[cfg(test)]
mod test_engine {
    use std::{fs, fs::File, io::Write};

    use insta::assert_debug_snapshot;
    use tempdir::TempDir;

    use super::*;

    const TEST_SENSITIVE_COMMANDS: &str = r###"
- name: Find me
  secret_group: 0
  test: FIND_ME=
    "###;

    const TEMP_HISTORY_LINES_CONTENT: &str = "history
ls
echo 'hello you'
rm -f ./file.txt
export FIND_ME=token
";

    const TEMP_HISTORY_FISH: &str = r#"---
- cmd: history
  when: "1656438759"
- cmd: ls
  when: "1656438760"
- cmd: echo 'hello you'
  when: "1656438760"
- cmd: rm -f ./file.txt
  when: "1656438760"
- cmd: export FIND_ME=token
  when: "1656438760"
"#;

    fn create_mock_state(
        temp_dir: &TempDir,
        content: &str,
        shell_type: shell::Shell,
    ) -> ShellContext {
        let app_folder = temp_dir.path().join("app");
        let history_file_name = "history";
        let history_file_path = app_folder.join(history_file_name);
        fs::create_dir_all(&app_folder).unwrap();

        let mut f = File::create(&history_file_path).unwrap();
        f.write_all(content.as_bytes()).unwrap();
        f.sync_all().unwrap();

        ShellContext {
            app_folder_path: app_folder.display().to_string(),
            history: shell::History {
                shell: shell_type,
                path: history_file_path.display().to_string(),
                file_name: history_file_name.to_string(),
            },
        }
    }

    #[test]
    fn can_find_history_commands_line() {
        let temp_dir = TempDir::new("engine").unwrap();

        let en = PatternsEngine {
            commands: serde_yaml::from_str(TEST_SENSITIVE_COMMANDS).unwrap(),
            masker: Masker::new(),
        };
        let state_context =
            create_mock_state(&temp_dir, TEMP_HISTORY_LINES_CONTENT, shell::Shell::Bash);

        let result = en.find_history_commands_from_shell_list(&vec![state_context]);

        assert_debug_snapshot!(result);
        temp_dir.close().unwrap();
    }

    // TODO: think of a better place for test that clear history

    #[test]
    fn can_clear_command_by_lines() {
        let temp_dir = TempDir::new("engine").unwrap();

        let en = PatternsEngine {
            commands: serde_yaml::from_str(TEST_SENSITIVE_COMMANDS).unwrap(),
            masker: Masker::new(),
        };
        let state_context =
            create_mock_state(&temp_dir, TEMP_HISTORY_LINES_CONTENT, shell::Shell::Bash);

        let result = en.find_history_commands_from_shell_list(&vec![state_context.clone()], true);

        assert_debug_snapshot!(result);
        assert_debug_snapshot!(fs::read_to_string(state_context.history.path));
        temp_dir.close().unwrap();
    }


    // TODO: think of a better place for test that clear history
    #[test]
    fn can_clear_find_fish() {
        let temp_dir = TempDir::new("engine").unwrap();

        let en = PatternsEngine {
            commands: serde_yaml::from_str(TEST_SENSITIVE_COMMANDS).unwrap(),
            masker: Masker::new(),
        };
        let state_context = create_mock_state(&temp_dir, TEMP_HISTORY_FISH, shell::Shell::Fish);

        let result = en.find_history_commands_from_shell_list(&vec![state_context.clone()], true);

        assert_debug_snapshot!(result);
        assert_debug_snapshot!(fs::read_to_string(state_context.history.path).unwrap());
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_find_history_commands_fish() {
        let temp_dir = TempDir::new("engine").unwrap();

        let en = PatternsEngine {
            commands: serde_yaml::from_str(TEST_SENSITIVE_COMMANDS).unwrap(),
            masker: Masker::new(),
        };
        let state_context = create_mock_state(&temp_dir, TEMP_HISTORY_FISH, shell::Shell::Fish);

        let result = en.find_history_commands_from_shell_list(&vec![state_context]);

        assert_debug_snapshot!(result);
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_find_custom_patterns() {
        let temp_dir = TempDir::new("engine").unwrap();

        let config = Config::with_custom_path(temp_dir.path().join("app"));
        config.init().unwrap();
        let custom_pattern = r###"
- name: Pattern Name
  test: FIND_ME
  secret_group: 1
  id: elad_ignore
"###;
        fs::write(&config.sensitive_commands_path, custom_pattern).unwrap();

        let en = PatternsEngine::with_config(&config).unwrap();
        let state_context =
            create_mock_state(&temp_dir, TEMP_HISTORY_LINES_CONTENT, shell::Shell::Bash);

        let result = en.find_history_commands_from_shell_list(&vec![state_context]);

        assert_debug_snapshot!(result);
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_ignore_patterns() {
        let temp_dir = TempDir::new("engine").unwrap();

        let config = Config::with_custom_path(temp_dir.path().join("app"));
        config.init().unwrap();
        let custom_pattern = r###"
- elad_ignore
"###;
        fs::write(&config.sensitive_commands_path, custom_pattern).unwrap();

        let en = PatternsEngine::with_config(&config).unwrap();
        let state_context =
            create_mock_state(&temp_dir, TEMP_HISTORY_LINES_CONTENT, shell::Shell::Bash);

        let result = en.find_history_commands_from_shell_list(&vec![state_context]);

        assert_debug_snapshot!(result);
        temp_dir.close().unwrap();
    }
}
