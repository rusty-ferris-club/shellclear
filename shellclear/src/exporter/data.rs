use anyhow::Result;
use chrono::{DateTime, NaiveDateTime, TimeZone, Utc};
use lazy_static::lazy_static;
use regex::Regex;

use crate::{
    data::FindingSensitiveCommands,
    shell::{FishHistory, Shell},
};

const DATE_TIME_FORMAT: &str = "%Y-%m-%d %H:%M:%S";
pub const LIMIT_COMMAND: usize = 100;

pub trait Exporter {
    /// export sensitive findings results
    ///
    /// # Errors
    ///
    /// Will return `Err` export has an error
    fn sensitive_data(&self, findings: &[FindingSensitiveCommands]) -> Result<()>;
}

lazy_static! {
    static ref ZSHRC_CAPTURE_COMMAND_TIME: Regex = Regex::new(r"^: ([0-9]+):").unwrap();
}

pub fn chunk(text: &str, size: usize) -> String {
    text.chars()
        .collect::<Vec<char>>()
        .chunks(size)
        .map(|c| c.iter().collect::<String>())
        .collect::<Vec<String>>()
        .join("\r\n")
}

/// Get command time execution
///
/// # Errors
///
/// Will return `Err` when conversion error
pub fn extract_time(finding: &FindingSensitiveCommands) -> Result<String> {
    match finding.shell_type {
        Shell::Zshrc => {
            if let Some(c) = ZSHRC_CAPTURE_COMMAND_TIME.captures(&finding.data) {
                if let Some(timestamp) = c.get(1) {
                    return Ok(format!(
                        "{}",
                        convert_str_timestamp_to_date_time(timestamp.as_str())?
                            .format(DATE_TIME_FORMAT)
                    ));
                }
            };
            Ok(finding.data.clone())
        }
        Shell::Fish => {
            let history: FishHistory = serde_yaml::from_str(&finding.data)?;
            Ok(format!(
                "{}",
                convert_str_timestamp_to_date_time(history.when.as_str())?.format(DATE_TIME_FORMAT)
            ))
        }
        _ => Ok("".to_string()),
    }
}

fn convert_str_timestamp_to_date_time(timestamp: &str) -> Result<DateTime<Utc>> {
    Ok(Utc.from_utc_datetime(&NaiveDateTime::from_timestamp(timestamp.parse::<i64>()?, 0)))
}

#[cfg(test)]
mod test_exporter {
    use std::str;

    use insta::assert_debug_snapshot;

    use super::*;
    use crate::shell::Shell;

    #[test]
    fn can_extract_time_zshrc() {
        let shell_finding = FindingSensitiveCommands {
            shell_type: Shell::Zshrc,
            sensitive_findings: vec![],
            command: "test command".to_string(),
            data: ": 1655110559:0;command data".to_string(),
            secrets: vec![],
        };

        let resp = extract_time(&shell_finding);

        assert_debug_snapshot!(resp);
    }

    #[test]
    fn can_extract_time_fish() {
        let shell_finding = FindingSensitiveCommands {
            shell_type: Shell::Fish,
            sensitive_findings: vec![],
            command: "test command".to_string(),
            data: r#"{ cmd: "export test command", when: "1655110559" }"#.to_string(),
            secrets: vec![],
        };

        let resp = extract_time(&shell_finding);

        assert_debug_snapshot!(resp);
    }
}
