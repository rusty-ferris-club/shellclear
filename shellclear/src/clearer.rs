use std::fmt::Write;
use std::fs::write;
use std::time::Instant;

use anyhow::Result;

use crate::engine::ShellCommands;
use crate::ShellContext;

pub struct Clearer {}

impl Clearer {
    /// Load sensitive ignores file
    ///
    /// # Errors
    ///
    /// Will return `Err` when the history file cannot be opened / written to
    pub fn write_findings(
        shells_context: &[ShellContext],
        commands: &ShellCommands,
        remove: bool,
    ) -> Result<()> {
        for context in shells_context {
            let start = Instant::now();
            let mut cleared_history: String = String::new();

            for command in commands
                .get_commands_per_shell(&context.history.shell)
                .unwrap_or(&vec![])
            {
                if remove {
                    if command.detections.is_empty() {
                        writeln!(&mut cleared_history, "{}", command.data)?;
                    }
                } else {
                    writeln!(&mut cleared_history, "{}", command.data)?;
                }
            }

            if !cleared_history.is_empty() {
                write(&context.history.path, cleared_history)?;
                log::debug!(
                    "time elapsed for backup existing file and write a new history to shell : {:?}",
                    start.elapsed()
                );
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::Path;

    use insta::assert_debug_snapshot;
    use regex::Regex;
    use tempdir::TempDir;

    use crate::data::{Command, Detection};
    use crate::shell::History;
    use crate::shell::Shell::Zshrc;

    use super::*;

    fn mock_state(dir: &Path) -> (ShellCommands, Vec<ShellContext>) {
        let state_context = vec![ShellContext {
            app_folder_path: "mock".to_string(),
            history: History {
                shell: Zshrc,
                file_name: "mock".to_string(),
                path: dir.to_str().unwrap().into(),
            },
        }];

        // We test one finding with a secret and one without a secret
        // The secret should be filtered out when using remove = true param
        let findings = vec![
            Command {
                command: "mock".to_string(),
                data: "mock".to_string(),
                detections: vec![],
                shell_type: Zshrc,
                secrets: vec![],
            },
            Command {
                command: "should be removed".to_string(),
                data: "should be removed".to_string(),
                detections: vec![Detection {
                    name: "mock".to_string(),
                    secret_group: 0,
                    id: "mock-id".to_string(),
                    test: Regex::new("").unwrap(),
                }],
                shell_type: Zshrc,
                secrets: vec!["mock".to_string()],
            },
        ];

        let mut commands = ShellCommands::default();

        commands.add_commands(&Zshrc, findings);

        (commands, state_context)
    }

    #[test]
    fn remove_sensitive_commands() {
        let dir = TempDir::new("clearer").unwrap();
        let history = dir.path().join("history");

        let (commands, state_context) = mock_state(&history);

        Clearer::write_findings(&state_context, &commands, true).unwrap();

        let content = fs::read_to_string(history).unwrap();

        assert_debug_snapshot!(content);

        dir.close().unwrap();
    }

    #[test]
    fn persist_sensitive_commands() {
        let dir = TempDir::new("clearer").unwrap();
        let history = dir.path().join("history");

        let (commands, state_context) = mock_state(&history);

        Clearer::write_findings(&state_context, &commands, false).unwrap();

        let content = fs::read_to_string(history).unwrap();

        assert_debug_snapshot!(content);

        dir.close().unwrap();
    }
}
