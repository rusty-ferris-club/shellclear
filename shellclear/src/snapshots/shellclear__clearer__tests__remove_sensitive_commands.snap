---
source: shellclear/src/clearer.rs
assertion_line: 131
expression: content
---
"mock\n"
